ping/ping_common.c

```diff
@@ -455,20 +455,26 @@ void sock_setbufs(struct ping_rts *rts, socket_st *sock, int alloc)
 	}
 }
 
-void sock_setmark(unsigned int mark, int fd)
+void sock_setmark(struct ping_rts *rts, int fd)
 {
 #ifdef SO_MARK
 	int ret;
 	int errno_save;
 
+	if (!rts->opt_mark)
+		return;
+
 	enable_capability_admin();
-	ret = setsockopt(fd, SOL_SOCKET, SO_MARK, &mark, sizeof(mark));
+	ret = setsockopt(fd, SOL_SOCKET, SO_MARK, &(rts->mark), sizeof(rts->mark));
 	errno_save = errno;
 	disable_capability_admin();
 
-	/* Do not exit, old kernels do not support mark. */
-	if (ret == -1)
-		error(0, errno_save, _("WARNING: failed to set mark: %u"), mark);
+	if (ret == -1) {
+		error(0, errno_save, _("WARNING: failed to set mark: %u"), rts->mark);
+		if (errno_save == EPERM)
+			error(0, 0, _("=> missing cap_net_admin+p capability?"));
```

Comment:
I guess I should check kernel version and print `cap_net_raw+p` only when relevant. I also wonder if print capability constants (`CAP_NET_RAW`, `CAP_NET_ADMIN`) instead of capability `setcap` capability (`cap_net_raw+p`, `cap_net_admin+p`).

Reply:
I could take LTP functions `tst_parse_kver()` and `parse_digit()` which parse `char release[]` member of `struct utsname`, to detect if kernel is new enough for `cap_net_raw+p` to be relevant. But maybe it's not worth to do that. So far simple `missing cap_net_admin+p or cap_net_raw+p (since Linux 5.17) capability?` without checking kernel version added.

https://github.com/linux-test-project/ltp/blob/db0082b45cc6f8f7c8c1c5f1cffb3ed5522d812e/lib/tst_kvercmp.c#L31-L82

Reply:
> I also wonder if print capability constants (CAP_NET_RAW, CAP_NET_ADMIN) instead of capability setcap capability (cap_net_raw+p, cap_net_admin+p).

I actually think the setcap is probably better here, since this is going to be seen by end users, and setcap is how the user would fix this.

I agree it probably isn't worth the effort to do version detection.  This LGTM.